models/remoteworkitem.go

```diff
@@ -0,0 +1,96 @@
+package models
+
+import (
+	"fmt"
+	"strconv"
+)
+
+const (
+	ProviderGithub string = "github"
+)
+
+// RemoteWorkItem represents a work item as it is stored in the database
+type RemoteWorkItem struct {
+	Fields map[string]interface{}
+}
+
+// WorkItemMap : holds the mapping between Remote key and local key in a WI
+type WorkItemMap map[string]interface{}
+
+// RemoteWorkItemSanitizer makes sure that the
+// Response from Provider API is sanitized into a
+// good looking flat 1-level dictionary
+type RemoteWorkItemMapper interface {
+	Flatten()
```

Comment:
> How does the RemoteWorkItem know everything the WorkItemMap needs is on the 1 level?

It does not, and when a key is not found, I am planning to handle that as an error. but your concern is right.

> Who is going to call this optional thing?

The integration code. Can there be a miss? Yes. I agree, not a great design.

> Who is making the decision to use GitHubRemoteWorkItem vs RemoteWorkItem?

If for a provider a derived type does not exist, then the onus is on the integration code to flatten and then call the `MapRemote(..)` method. Else, the integration code can invoke a correct derivative of `RemoteWorkItem`. 

This is what I intended to do initially:

I would have wanted to have every derivative of  `RemoteWorkItem` override the`Flatten()` method, and at call the same from `func (rwi *RemoteWorkItem) MapRemote` .  
But I don't see how I can do that in Go.